### PR TITLE
Add test coverage for readme examples

### PR DESCRIPTION
**Pull request by davisshaver**
See https://github.com/fusioneng/speed-bumps/issues/53

### CODE REVIEW

**Comment by davisshaver** [comment]
@goldenapples review/merge on this basic filter usage test? I'm going to use it as the basis for readme update.

File: tests/test-speed-bumps-filter-usage.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+		$this->speed_bumps->register_speed_bump( 'speed_bump_test', array(
+			'string_to_inject' => function() { return '<div id="speed-bump-test"></div>'; },
+			'minimum_content_length' => false,
+			'from_start' => false,
+			'from_end' => false,
+		));
+
+		add_filter( 'the_content', function( $content ) {
+			return apply_filters( 'speed_bumps_inject_content', $content );
+		}, 1 );
```
Comment (goldenapples) [inline tests/test-speed-bumps-filter-usage.php]: I really prefer passing a function name in to the filter.

```
add_filter( 'the_content', array( Speed_Bumps(), 'insert_speed_bumps' ) );
```

would work here. Looks like that function has been renamed since the initial readme doc was written.

File: tests/test-speed-bumps-filter-usage.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+And then, I was gone.
+
+While GIFs may seem like a flash in the pan—really, how can four seconds turn you on?—the nature of the loop actually allows the viewer to spend an elongated amount of time taking in the presented scenario. GIFs give the viewer time to notice the caress of a hand floating from neck to shoulder to forearm, the tensing of an abdomen, the arching of a back, and the reflex of a thigh. After a few loops, you may find yourself empathizing with the players involved. Maybe you can even feel what they’re feeling.
+
+Of course, porn GIFs don’t appeal only to women, but the “microporn” does appear to have struck a unique chord with the ladies, according to Hester—an audience known to feel alienated by mainstream porn, historically geared toward men. Spend five minutes on Tumblr, and you’ll find yourself sucked into a pulsing subculture of porn GIFs curated for women, living on pages like Porn-Gifs-For-Women and YummyPornForGirls. GIFs that almost exclusively spotlight erotic female pleasure.';
```
Comment (goldenapples) [inline tests/test-speed-bumps-filter-usage.php]: nice example text :point_right: :ok_hand:

**Comment by goldenapples** [comment]
Good call to have any functions referenced in documentation covered by tests. I'd like to give people a simpler filter example without requiring them to define a closure function just to pass one filter to another. 

Would it be good to expose a function outside of the Speed_Bumps class to make this process even simpler?

File: tests/test-speed-bumps-filter-usage.php (hunk 2 above)
Comment (davisshaver) [inline tests/test-speed-bumps-filter-usage.php]: Source for posterity http://fusion.net/story/165548/why-women-love-porn-gifs/

**Comment by davisshaver** [comment]
@goldenapples Updated with simpler filter call. 

As for breaking `register_speed_bump` out of the class, having a global equivalent would be a bit more natural for the plugin (pattern I see usually is leaving the global functions as the main I/O for integrating the plugin). Do we have any similar patterns in existing projects? I can take a stab at the wrapper (I imagine we'd just be calling `Speed_Bumps()->register_speed_bump` internally?)

**Comment by goldenapples** [comment]
> Do we have any similar patterns in existing projects?

I tried to do something like this in [Co-Authors Plus Roles](https://github.com/fusioneng/Co-Authors-Plus-Roles/blob//co-authors-plus-roles.php#L82-100). I'm not sure I would hold up that plugin as a model of good architecture, but I do like the concept of keeping all common public methods in globals and defining them all in one place for easy reference.

**Comment by davisshaver** [comment]
@goldenapples Sweet, that's what I had in mind. I'll try adding a register function and test here.

**Comment by davisshaver** [comment]
@goldenapples Should we expose any other methods than `register_speed_bump`?

**Comment by goldenapples** [comment]
Maybe `insert_speed_bumps`, so that it can be used as a filter on `the_content`?

**Comment by davisshaver** [comment]
:100:. Minor code change, but feels a lot easier to use because of it.

**Comment by davisshaver** [comment]
@goldenapples How do you feel about this? cc @noppanit

**Comment by goldenapples** [comment]
I like it a lot more. We might want to get @danielbachhuber's take on adding functions to the global namespace, but I think that exposing this kind of basic functionality in the global namespace is a big win as far as readability.

Do we even need the `speed_bumps_inject_content` filter any more? It seems like this might be a good time to remove it and direct people towards these better ways of filtering the content.

File: speed-bumps.php [hunk 3]
```diff
@@ -260,3 +260,17 @@ function Speed_Bumps() {
...
+ *
+ * All functions that should be available in the global namespace are listed here.
+ * This would be done with `use` statements in PHP 5.6+.
+ *
+ */
+function register_speed_bump( $id, $args = array() ) {
+	return Speed_Bumps()->register_speed_bump( $id, $args );
+}
+
+function insert_speed_bumps( $thecontent ) {
```
Comment (danielbachhuber) [inline speed-bumps.php]: What's the need to register these functions?

Comment (goldenapples) [inline speed-bumps.php]: IHO - doing this clearly indicates to users which parts of the plugin functionality you expect them to interact with. It also makes common functions a little clearer to read:

_instead of:_
```
add_filter( 'the_content', array( Speed_Bumps(), 'insert_speed_bumps' ) );
```

_you can just write:_
```
add_filter( 'the_content', 'insert_speed_bumps' );
```

Personal preference only, but I like this pattern for exposing common use cases.

File: speed-bumps.php [hunk 4]
```diff
@@ -260,3 +260,17 @@ function Speed_Bumps() {
 
 add_action( 'init', 'Speed_Bumps' );
 
+/**
+ * The Public API for this plugin.
+ *
+ * All functions that should be available in the global namespace are listed here.
+ * This would be done with `use` statements in PHP 5.6+.
```
Comment (goldenapples) [inline speed-bumps.php]: I don't remember what I was thinking when I initially wrote this comment, but I don't see now how `use` statements would help here. Can we remove this line?

File: speed-bumps.php (hunk 3 above)
Comment (davisshaver) [inline speed-bumps.php]: @danielbachhuber @goldenapples After setting up in a local installation, I feel the global functions make theme integration easier to read. As @goldenapples mentioned, it also makes explicit how we expect the developer to integrate the plugin (vs what is "core" code).

One downside is that the addition of this code (in either instance) causes the theme to fatal if the plugin isn't active, but I'm not sure of a clean way around that given current registration architecture. We could recommend that it be added within its own plugin file, I guess that would cause the fatal to deactivate the registration.

**Comment by davisshaver** [comment]
@goldenapples No worries. So I'd say that we need to fix https://github.com/fusioneng/speed-bumps/issues/11 before the release then – agree/disagree? I think both tests should function the same from the end user perspective. I'll plan to pick it up first thing Monday if you think this i the right direction to go.

**Comment by goldenapples** [comment]
> So I'd say that we need to fix #11 before the release then – agree/disagree?

If you'd like to give that issue a shot, I'd say go for it. I have some fear that trying to implement passable HTML parsing in this context might be too severe of a performance hit to be workable. (#37) It also adds a lot of complication, in that what is currently a simple loop through possible insertion points would become a recursive tree crawl with more heavy DOM parsing to build the context necessary for evaluating each iteration. 

We could also just specify the priority to attach the filter with, as is done in the inline documentation: https://github.com/fusioneng/speed-bumps/blob/master/speed-bumps.php#L103-105

**Comment by davisshaver** [comment]
@goldenapples After re-reading this thread, I see what you mean regarding priority on the filter. That seems like a more sensible move for the scope of this pull request. Sounds like #11 is a more involved effort (with more potential breakage) than makes sense for the v0.1.0 release.

What do you think of the current test coverage? See any issues?

I'm currently trying to debug why this snippet fails to add the speed bump in my local environment:
```
	register_speed_bump( 'needy_rickroll', array(
		'string_to_inject' => function() { return '<iframe width="420" height="315" src="https://www.youtube.com/embed/dQw4w9WgXcQ" frameborder="0" allowfullscreen></iframe>'; },
		'minimum_content_length' => false,
		'from_start' => array(
			'paragraphs' => 2,
		),
		'from_end' => false,
	));
```

**Comment by goldenapples** [comment]
> What do you think of the current test coverage? See any issues?

Very nice. Setting up special coverage for everything we refer to in documentation is a great idea. Overall, I think the general structure of our test coverage is decent, but it could definitely stand better organization still. I'd also like to have test coverage for some of the ideas that are sketched out, like defining new element constraints. I defined a few new elements [in our theme](https://github.com/fusioneng/fusion-theme/blob/master/inc/integrations/speed-bumps/constraints.php) and wrote tests for them [there](https://github.com/fusioneng/fusion-theme/blob/master/php-tests/integrations/test-speed-bumps.php#L146-253), but I realized that functionality isn't covered by tests in the plugin.

> I'm currently trying to debug why this snippet fails to add the speed bump in my local environment.

Hmm. What content are you trying to insert it in? What's the difference between it and the passing test you have in there? The only difference I see is the passing test has a value set for `minimum_content_length`... is there a bug there?

**Comment by davisshaver** [comment]
@goldenapples yes, after working on this readme, I think testing all doc code should become a regular habit for the team :).

As for the filter issue, still trying to debug. I think part of the confusion came from the changing arg structure, and letting my clean WP server fall behind master. Working through that now.

Regarding 
> I'd also like to have test coverage for some of the ideas that are sketched out, like defining new element constraints. I defined a few new elements in our theme and wrote tests for them there, but I realized that functionality isn't covered by tests in the plugin.

Do you feel these should be included in v0.1.0 release? Or should we start stubbing these to v0.2.0 where applicable?

**Comment by davisshaver** [comment]
So, here is the problem I'm trying to work through.

I've added a real simple speed bump to 2015.

```
register_speed_bump( 'speed_bump_sample', array(
	'string_to_inject' => function() { return '<div id="speed-bump-sample"></div>'; },
));
add_filter( 'the_content', 'insert_speed_bumps', 1 );
```

I want to document where I expect the speedbump to get placed. Let's see – after the 75th word, or after the 3rd paragraph (based on https://github.com/fusioneng/speed-bumps/blob/master/speed-bumps.php#L172-L180), whichever comes later, as long as it's more than 3 paragraphs from the end of the article.

Based on this logic, I've marked where I expect the speedbump to go. However you will see inline where it got placed. 

```
<p>Far far away, behind the word mountains, far from the countries Vokalia and Consonantia, there live the blind texts.</p>
<p>Separated they live in Bookmarksgrove right at the coast of the Semantics, a large language ocean.</p>
<p>A small river named Duden flows by their place and supplies it with the necessary regelialia. It is a paradisematic country, in which roasted parts of sentences fly into your mouth.</p>
<p>Even the all-powerful Pointing has no control about the blind texts it is an almost unorthographic life One day however a small line of blind text by the name of Lorem Ipsum decided to leave for the far World of Grammar.</p>
*EXPECTED HERE*
<p>The Big Oxmox advised her not to do so, because there were thousands of bad Commas, wild Question Marks and devious Semikoli, but the Little Blind Text didn’t listen.</p>
<div id="speed-bump-sample"></div>
<p>She packed her seven versalia, put her initial into the belt and made herself on the way.</p>
<p>When she reached the first hills of the Italic Mountains, she had a last view back on the skyline of her hometown Bookmarksgrove, the headline of Alphabet Village and the subline of her own road, the Line Lane.</p>
<p>Pityful a rethoric question ran over her cheek, then she continued her way. On her way she met a copy. The copy warned the Little Blind Text, that where it came from it would have been rewritten a thousand times and everything that was left from its origin would be the word &#8220;and&#8221; and the Little Blind Text should</p>
```

My goal is to demonstrate a simple use of the `from_start` parameter. I would like to show where the default placement is, and how the placement can be modified with a code change.

**Comment by goldenapples** [comment]
> I think testing all doc code should become a regular habit for the team

:+1: yup. I'll make much more of an effort to actually set up a test before suggesting that something should work or writing documentation of a feature...

**Comment by davisshaver** [comment]
In the readme, I'm now demonstrating how placement can be modified with this line. I'm trying to figure out what I misunderstand regarding placement logic, as described above. @goldenapples see anything I'm missing?

Anyway here is the sample I'm including to demonstrated earlier placement within the content.

```
register_speed_bump( 'speed_bump_sample', array(
	'string_to_inject' => function() { return '<div id="speed-bump-sample"></div>'; },
	'from_start' => 0,
));
```

As expected, this results in the speed bump being injected after the first (or, 0th) paragraph. So this meets expectations, but I would still like to resolve what I'm misinterpreting the above comment. 

```
<p>Far far away, behind the word mountains, far from the countries Vokalia and Consonantia, there live the blind texts.</p>
<div id="speed-bump-sample"></div>
<p>Separated they live in Bookmarksgrove right at the coast of the Semantics, a large language ocean.</p>
<p>A small river named Duden flows by their place and supplies it with the necessary regelialia. It is a paradisematic country, in which roasted parts of sentences fly into your mouth.</p>
<p>Even the all-powerful Pointing has no control about the blind texts it is an almost unorthographic life One day however a small line of blind text by the name of Lorem Ipsum decided to leave for the far World of Grammar.</p>
<p>The Big Oxmox advised her not to do so, because there were thousands of bad Commas, wild Question Marks and devious Semikoli, but the Little Blind Text didn’t listen.</p>
<p>She packed her seven versalia, put her initial into the belt and made herself on the way.</p>
<p>When she reached the first hills of the Italic Mountains, she had a last view back on the skyline of her hometown Bookmarksgrove, the headline of Alphabet Village and the subline of her own road, the Line Lane.</p>
<p>Pityful a rethoric question ran over her cheek, then she continued her way. On her way she met a copy. The copy warned the Little Blind Text, that where it came from it would have been rewritten a thousand times and everything that was left from its origin would be the word &#8220;and&#8221; and the Little Blind Text should</p>
```

**Comment by goldenapples** [comment]
Oh, looks like you're seeing a bug in the `meets_minimum_distance_from_start` function.

I see where the issue is: https://github.com/fusioneng/speed-bumps/blob/add-test-for-basic-speed-bump-filter/inc/constraints/text/class-minimum-text.php#L72

An array slice with a start point of 0 and length of the current index is going to be off by one in terms of how many paragraphs it actually counts. Simple stupid bug on my part. I need to check into why none of the tests caught it...

**Comment by davisshaver** [comment]
:bug: !!

**Comment by davisshaver** [comment]
@goldenapples lemme know if you want test or pull to fix that, or I'll watch for yours. After we get this cleared up, I think we'll be in good shape to release. I'll get started on the post in meantime.

**Comment by goldenapples** [comment]
The single hardest problem in coding is good test coverage and off by one errors.

I'm heading out for the day; I can put together a pull request in the morning; or if you're on a roll, feel free to go ahead in the meantime. Looks like the same bug is going to affect the from_end parameter too.

**Comment by davisshaver** [comment]
sweet! Behavior on the sample filter is now exactly as expected (e.g. speed bump placed where I had expected it in https://github.com/fusioneng/speed-bumps/pull/54#issuecomment-123042327).

The readme and test coverage should now be in parity.

One shortcoming of existing test coverage on the filter usage: `assertSpeedBumpAtParagraph` relies on our "two linebreaks == paragraph" heurestic. However, by the time the test runs (following `apply_filters( 'the_content', $post->post_content )`, the content contains `<p>` tags, so we'll need another explosion function to identify paragraph placement for filtered content.

As @goldenapples noted in https://github.com/fusioneng/speed-bumps/issues/60, we'll still need to think/work through the larger index consistency issue described in https://github.com/fusioneng/speed-bumps/issues/52, but that can come with a later version.

**Comment by davisshaver** [comment]
Algorithm question @noppanit @goldenapples … 

I want to use `assertSpeedBumpAtParagraph` on content that already has `<p>` tags instead of double linebreaks (which is what  `assertSpeedBumpAtParagraph` currently uses as its "paragraph" heuristic https://github.com/fusioneng/speed-bumps/blob/d15de4e760bc8fee6a92f63e90aa08e8ce016dec/tests/test-speed-bumps-integration.php#L190-L200). 

Right now we don't have a good way to measure speed bump placement on filtered content. Current test coverage uses `insert_speed_bumps` directly instead of as a filter on `the_content`.

Is there an easy way explode into paragraph parts? or maybe we should filter the `<p>` back to `\n\n` on the fly?

**Comment by noppanit** [comment]
Oh that's a good point. @goldenapples can correct me if I'm wrong. There's a filter in WordPress https://codex.wordpress.org/Function_Reference/wpautop to add `<p>` but we add speed bumps before that point so the content won't have `<p>`. We can make a new method to test the content with `<p>`.

I'm not an expert in Regex but I think it's better to use `DOMDocument` if we want to extract `<p`> and less error-prone.

I'd imagine it would be something like

```
$dom = new DOMDocument();
$dom->loadHTML($a);
$domx = new DOMXPath($dom);
$entries = $domx->evaluate("//p");
$parts = array();
foreach ($entries as $entry) {
    $parts[] = $entry->nodeValue;
}

```

**Comment by goldenapples** [comment]
Yes, that's the issue I was trying to get around by recommending that, if people attach `insert_speed_bumps` as a filter on `the_content`, they do it at a very low priority so as to get access to the content before all of WP's filters have happened.
If you want to test our filter on `the_content`, but remove all of WordPress's default filters, you might want to just remove the default filters before running the test:

```
remove_all_filters( 'the_content', 10 );
```
You may also want to remove all the filters at priority `8` (autoembed and run_shortcode) and `11` (do_shortcode), depending on what you want to test for. `10` is where all the default texturizing filters are attached.

**Comment by davisshaver** [comment]
@noppanit Thanks for the `DOMDocument` pointer. Unfortunately I ran into issues with the `<iframe>` and testing its presence, haven't dug into it but appears that the string gets stripped by the time `nodeValue` is called.

@goldenapples, I tried to implement per your comment. Would you mind giving me a review of this test coverage? 

This new test coverage should match up with code in the readme https://github.com/fusioneng/speed-bumps/blob/improve-readme-for-first-release/README.md, which is also ready for review/merge.

File: tests/test-speed-bumps-filter-usage.php [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+	public function test_speed_bump_filter_usage() {
+		/* Removing all default filters so we can test placement */
+		remove_all_filters( 'the_content', 10 );
+		register_speed_bump( 'speed_bump_test', array(
+			'string_to_inject' => function() { return '<div id="speed-bump-test"></div>'; },
+		));
+		$post_id = $this->factory->post->create( array( 'post_content' => $this->get_dummy_content() ) );
+		$post = get_post( $post_id );
+		$this->assertSpeedBumpAtParagraph( apply_filters( 'the_content', $post->post_content ), 5, '<div id="speed-bump-test"></div>' );
```
Comment (goldenapples) [inline tests/test-speed-bumps-filter-usage.php]: I'm not sure that creating a post here is actually necessary. This does exactly the same thing as if you just ran `apply_filters( 'the_content' ` over the dummy content. Slowing down the test suite by having each test hit the database feels like overkill to me.

File: tests/test-speed-bumps-filter-usage.php [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+		$rick_roll_post = get_post( $rick_roll_id );
+
+		add_filter( 'speed_bumps_rickroll_constraints', 'give_you_up', 10, 4 );
+
+		function give_you_up( $can_insert, $context, $args, $already_inserted ) {
+			if ( ! preg_match( '/give [^ ]+ up/i', $context['prev_paragraph'] ) ) {
+				$can_insert = false;
+			}
+			return $can_insert;
+		}
```
Comment (goldenapples) [inline tests/test-speed-bumps-filter-usage.php]: Nice test. I don't think there was anything in the existing test suite that covered adding an additional constraint to a specific speed bump like this.

File: tests/test-speed-bumps-filter-usage.php [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+			'from_end' => false,
+		) );
+
+		$non_rick_roll_id = $this->factory->post->create( array( 'post_content' => $this->get_dummy_content() ) );
+		$non_rick_roll_post = get_post( $non_rick_roll_id );
+
+		$rick_roll_id = $this->factory->post->create( array( 'post_content' => $this->get_rick_rolly_content() ) );
+		$rick_roll_post = get_post( $non_rick_roll_id );
+
+		add_filter( 'speed_bumps_rickroll_constraints', '__return_false', 11 );
```
Comment (goldenapples) [inline tests/test-speed-bumps-filter-usage.php]: Why did you need to use priority 11? All of the default filters *should* return false if false was passed in to them. 

It might be good to also add a test where you add a falsy filter at priority 9 and ensure that none of the default filters change that false to true.

File: tests/test-speed-bumps-filter-usage.php (hunk 6 above)
Comment (goldenapples) [inline tests/test-speed-bumps-filter-usage.php]: Would a more specific name for the test be good, so that we can see in the future what this test was actually covering? `test_rickroll_example_with_two_filters` describes its role in the documentation, but not what actual functionality it covers.

**Comment by goldenapples** [comment]
I think it's looking good. Left a couple comments. If they make sense to address, then we can merge and release soon...

File: tests/test-speed-bumps-filter-usage.php (hunk 7 above)
Comment (davisshaver) [inline tests/test-speed-bumps-filter-usage.php]: @goldenapples Updated the test to omit priority.

> It might be good to also add a test where you add a falsy filter at priority 9 and ensure that none of the default filters change that false to true.
Cool, just did this. Would appreciate a sanity check though – I believe what we're testing is that if one constraint says "No!", make sure that none of the default filters override it.

File: tests/test-speed-bumps-filter-usage.php (hunk 6 above)
Comment (davisshaver) [inline tests/test-speed-bumps-filter-usage.php]: @goldenapples I'd like to do two things
- Rename this test file to state that its explicitly to test readme code.
- Add another test in the integrations test file to test the constraint.

Proceeding that way now...

File: tests/test-speed-bumps-filter-usage.php (hunk 7 above)
Comment (goldenapples) [inline tests/test-speed-bumps-filter-usage.php]: > I believe what we're testing is that if one constraint says "No!", make sure that none of the default filters override it.

Exactly. There should be nothing in the default filters that returns `true` if an earlier filter has returned `false`, and its good to have a test covering that.

File: tests/test-speed-bumps-filter-usage.php (hunk 6 above)
Comment (goldenapples) [inline tests/test-speed-bumps-filter-usage.php]: :+1:

**Comment by davisshaver** [comment]
@goldenapples second look?

**Comment by goldenapples** [comment]
LGTM. Is this good to merge?

**Comment by davisshaver** [comment]
@goldenapples Good to go from my end.
